datapath/iptables: Check iptables kernel modules

vadorovsky opened the pull request:
Before this change, ip6tables was called for rules removal even if
ip6tables modules were not loaded.

After this change, iptables module is going to:
- fail if iptables modules are not loaded
- fail if ip6tables modules are not loaded, but IPv6 is enabled
- not use ip6tables if its modules are not loaded and IPv6 is disabled

Fixes: #7453
Fixes: 65cfe4a ("iptables: Add support for IPv6, make IPv4 optional.")
Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7464)
<!-- Reviewable:end -->

vadorovsky committed:
datapath/iptables: Check iptables kernel modules

Before this change, ip6tables was called for rules removal even if
ip6tables modules were not loaded.

After this change, iptables module is going to:
- fail if iptables modules are not loaded
- fail if ip6tables modules are not loaded, but IPv6 is enabled
- not use ip6tables if its modules are not loaded and IPv6 is disabled

Fixes: #7453
Fixes: 65cfe4a6cf49 ("iptables: Add support for IPv6, make IPv4 optional.")
Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

aanm reviewed (changes requested):
It's not this simple @mrostecki. If the user was running with IPv6 only and switches back to IPv4 only, the ip6tables rules are not removed. (The same goes if the users was running a IPv4 only and switches to IPv6 only)

Probably it would be best to check if we can run `ip6tables` successfully before trying to delete the rules?

vadorovsky committed:
set: Add utility for subset checks

This package provides a function which checks whether the first slice of
strings is a subset of the second one. It's going for searching loaded
kernel modules.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
modules: Add utility for checking loaded kernel modules

This package provides a manager for checking and searching loaded kernel
modules. It's going to be used for checking whether iptables and
ip6tables are usable.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

tgraf commented on pkg/modules/modules_test.go:
```diff
@@ -0,0 +1,166 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !privileged_tests
```
I was expecting this to be a privileged test. Can we really rely on modprobe without privileges?

vadorovsky replied:
This file does not test the `FindOrLoadModules` function which calls modprobe. I will write a separate privilege test for that.

vadorovsky replied:
Done.

tgraf commented on pkg/modules/modules.go:
```diff
@@ -0,0 +1,63 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package modules
```
Please move this into `pkg/datapath/linux` as we want to isolate all new Linux specific code

vadorovsky replied:
1. This package handles only kernel modules listing and loading. I don't think that keeping it inside `pkg/datapath` package makes sense, because it does not relate to datapath itself.
2. What about moving all Linux-specific stuff to `modules_linux.go` and trying to keep `modules.go` OS-agnostic? I guess that anyone who wants to support the other platforms could just create i.e. a `modules_darwin.go` or `modules_freebsd.go` module.

ianvernon replied:
@tgraf ^

vadorovsky replied:
I moved the `modprobe` usage to `modules_linux.go` and privileged unit test to `modules_linux_privileged_test` with a `linux` build constraint, so now that module should be independent from platform. Developing any other `modules_<platform>.go` module which implements `moduleLoader` function (returning the name of the tool, like i.e. `kldload` in FreeBSD) should be possible now, hopefully without touching the main module. Please let me know whether you agree with that solution.

tgraf commented on pkg/set/set.go:
```diff
@@ -0,0 +1,41 @@
+// Subset checks whether the first slice is a subset of the second slice. If
+// not, it also returns slice of elements which are the difference of both
+// input slices.
+func Subset(sub, main []string) (bool, []string) {
```
Let's call this `SliceSubsetOf()`

tgraf commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -214,28 +215,62 @@ var ciliumChains = []customChain{
+	if err := modulesManager.FindOrLoadModules(
+		"ip_tables", "iptable_nat", "iptable_mangle", "iptable_raw",
+		"iptable_filter"); err != nil {
+		log.WithError(err).Fatal("iptables kernel could not be loaded")
```
Change this to `iptables modules could not be initialized`

tgraf commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -214,28 +215,62 @@ var ciliumChains = []customChain{
+		"ip6_tables", "ip6table_mangle", "ip6table_raw"); err != nil {
+		if option.Config.EnableIPv6 {
+			log.WithError(err).Fatal(
+				"IPv6 is enabled, but ip6tables kernel modules are not loaded")
```
`IPv6 is enabled and ip6tables modules could not be initialized`

tgraf reviewed (changes requested):
This doesn't address the IPv6 only case but that's not a regression in this PR but was already incomplete before. Overall looks good, some code organization changes that would make sense.

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -214,28 +215,62 @@ var ciliumChains = []customChain{
+			"ip6tables kernel modules could not be loaded, so IPv6 cannot be used")
+		ip6tables = false
+	}
+	m.ip6tables = ip6tables
```
If ipv6 is disabled in the config, then we will still try to manage/delete those ipv6 rules so long as the modules are loaded correctly. Is this right, or am I missing something?

vadorovsky replied:
Yes, this is right. The `ip6tables` field is going to be `true` every time the ip6tables modules are available. It's only false when those modules are not available to load.

joestringer replied:
This seems a bit counter-intuitive, I'm wondering why would we manage ip6tables rules if ipv6 is disabled in Cilium?

vadorovsky replied:
Because someone might decide to change the configuration and restart Cilium. For example, first have Cilium with ipv6 enabled and then change it to disabled. In such a case I think we should remove/cleanup ip6tables rules.
